src/pybind/mgr/volumes/fs/operations/access.py

```diff
@@ -0,0 +1,126 @@
+import errno
+import json
+
+def allow_access(mgr, client_entity, want_mds_cap, want_osd_cap,
+               unwanted_mds_cap, unwanted_osd_cap):
+    ret, out, err = mgr.mon_command({
+        "prefix": "auth get",
+        "entity": client_entity,
+        "format": "json"})
+
+    if ret == -errno.ENOENT:
+        ret, out, err = mgr.mon_command({
+            "prefix": "auth get-or-create",
+            "entity": client_entity,
+            "caps": ['mds',  want_mds_cap, 'osd', want_osd_cap, 'mon', 'allow r'],
+            "format": "json"})
+    else:
+        cap = json.loads(out)[0]
+
+        def cap_update(
+                orig_mds_caps, orig_osd_caps, want_mds_cap,
+                want_osd_cap, unwanted_mds_cap, unwanted_osd_cap):
+
+            if not orig_mds_caps:
+                return want_mds_cap, want_osd_cap
+
+            mds_cap_tokens = orig_mds_caps.split(",")
+            osd_cap_tokens = orig_osd_caps.split(",")
+
+            if want_mds_cap in mds_cap_tokens:
+                return orig_mds_caps, orig_osd_caps
+
+            if unwanted_mds_cap in mds_cap_tokens:
+                mds_cap_tokens.remove(unwanted_mds_cap)
+                osd_cap_tokens.remove(unwanted_osd_cap)
+
+            mds_cap_tokens.append(want_mds_cap)
+            osd_cap_tokens.append(want_osd_cap)
+
+            return ",".join(mds_cap_tokens), ",".join(osd_cap_tokens)
+
+        orig_mds_caps = cap['caps'].get('mds', "")
+        orig_osd_caps = cap['caps'].get('osd', "")
+
+        mds_cap_str, osd_cap_str = cap_update(
+            orig_mds_caps, orig_osd_caps, want_mds_cap, want_osd_cap,
+            unwanted_mds_cap, unwanted_osd_cap)
+
+        mgr.mon_command(
+            {
+                "prefix": "auth caps",
+                'entity': client_entity,
+                'caps': [
+                    'mds', mds_cap_str,
+                    'osd', osd_cap_str,
+                    'mon', cap['caps'].get('mon', 'allow r')],
+            })
+        ret, out, err = mgr.mon_command(
+            {
+                'prefix': 'auth get',
+                'entity': client_entity,
+                'format': 'json'
+            })
+
+    # Result expected like this:
+    # [
+    #     {
+    #         "entity": "client.foobar",
+    #         "key": "AQBY0\/pViX\/wBBAAUpPs9swy7rey1qPhzmDVGQ==",
+    #         "caps": {
+    #             "mds": "allow *",
+    #             "mon": "allow *"
+    #         }
+    #     }
+    # ]
+
+    caps = json.loads(out)
+    assert len(caps) == 1
+    assert caps[0]['entity'] == client_entity
+    return caps[0]['key']
+
+def deny_access(mgr, client_entity, want_mds_caps, want_osd_caps):
```

Comment:
I thought it was considered necessary to save the authorized clients somewhere so that we can check for it in deauthorize, here. Otherwise you can delete any entity, yes?

Reply:
A list of clients authorized to access a subvolume is stored in a file by ceph_volume_client to quickly fetch the authorized list required by manila to make bulk access rule changes. The authorized list was not maintained to check whether the client can be de-authorized by the ceph_volume_client.

Reply:
Is it no longer necessary to make "bulk access rule changes"? What kind of changes?

I think it makes sense for the volumes plugin to keep track of what it's authorized. If a volume is deleted, we should also remove the authorizations.

Reply:
> Is it no longer necessary to make "bulk access rule changes"? What kind of changes?
>

Yes, we still need to enable manila to do bulk access rule changes, e.g., delete all rules associated with a share when deleting or unmanaging a share, https://github.com/openstack/manila/blob/master/manila/share/manager.py#L3169
I had created a separate tracker issue, https://tracker.ceph.com/issues/44931 for the feature to keep track of auth IDs that have access to a subvolume/share. I wanted to break the manila allow/deny requirement into reasonable chunks. Besides this tracker ticket, we also need to implement https://tracker.ceph.com/issues/44928 for manila.

manila driver has an update_access API that requires the storage system to fetch all the the authorized entities (IPs, cephx auth IDs, etc) that have been granted access to a share (subvolume). Since ceph natively has an interface only to list subvolumes that have access to a cephx auth ID (using ceph auth get command), and not the other way around as required for manila, we have a mechanism in ceph_volume_client to store a list of auth IDs that have access to a subvolume in a per subvolume meta file.

> I think it makes sense for the volumes plugin to keep track of what it's authorized. If a volume is deleted, we should also remove the authorizations.

Did you mean that we should remove authorizations when deleting a subvolume (and not a volume)?

At least with ceph_volume_client we don't worry about cleaning up authorizations granted for a subvolume as manila does this for us. When you delete or unmanage a manila share, the manila share service issues calls to the driver to delete all the access rules associated with the share. See,
https://github.com/openstack/manila/blob/master/manila/share/manager.py#L3169 

Are you suggesting that during subvolume deletion mgr/volumes should be able to clean up authorizations associated with the subvolume outside of manila use-case? In that case, yeah, we should look up the list of auth IDs that have access to the subvolume stored in a meta file, and then call subvolume.deauthorize() for each of the auth IDs.

To remove subvolume authorizations during volume deletion, we'd have to iterate through the subvolume meta files and deauthorize the auth IDs?  It'd be very inefficient.

Reply:
> > Is it no longer necessary to make "bulk access rule changes"? What kind of changes?
> 
> Yes, we still need to enable manila to do bulk access rule changes, e.g., delete all rules associated with a share when deleting or unmanaging a share, https://github.com/openstack/manila/blob/master/manila/share/manager.py#L3169
> I had created a separate tracker issue, https://tracker.ceph.com/issues/44931 for the feature to keep track of auth IDs that have access to a subvolume/share. I wanted to break the manila allow/deny requirement into reasonable chunks. Besides this tracker ticket, we also need to implement https://tracker.ceph.com/issues/44928 for manila.

Alright, that division of the feature into tickets is okay with me.

> manila driver has an update_access API that requires the storage system to fetch all the the authorized entities (IPs, cephx auth IDs, etc) that have been granted access to a share (subvolume). Since ceph natively has an interface only to list subvolumes that have access to a cephx auth ID (using ceph auth get command), and not the other way around as required for manila, we have a mechanism in ceph_volume_client to store a list of auth IDs that have access to a subvolume in a per subvolume meta file.
> 
> > I think it makes sense for the volumes plugin to keep track of what it's authorized. If a volume is deleted, we should also remove the authorizations.
> 
> Did you mean that we should remove authorizations when deleting a subvolume (and not a volume)?

Yes, but maybe we should think about the same for volumes as well.

> At least with ceph_volume_client we don't worry about cleaning up authorizations granted for a subvolume as manila does this for us. When you delete or unmanage a manila share, the manila share service issues calls to the driver to delete all the access rules associated with the share. See,
> https://github.com/openstack/manila/blob/master/manila/share/manager.py#L3169

Even so, I think we should go ahead and delete any authorizations by default even if Manila automates that. It will use the same code paths I think.

> Are you suggesting that during subvolume deletion mgr/volumes should be able to clean up authorizations associated with the subvolume outside of manila use-case?

Yes. We should generally clean up after ourselves.

> In that case, yeah, we should look up the list of auth IDs that have access to the subvolume stored in the subvolume meta file, and then call subvolume.deauthorize() for each of the auth IDs.
> https://github.com/ceph/ceph/blob/master/src/pybind/ceph_volume_client.py#L913, fetching subvolume metadata file contents
> https://github.com/ceph/ceph/blob/master/qa/tasks/cephfs/test_volume_client.py#L763, shows contents of a subvolume meta file.
> 
> To remove subvolume authorizations during volume deletion, we'd have to iterate through the auth meta files and make authorization changes accordingly. Currently, ceph_volume_client stores auth meta files for all the auth IDs that have been granted access via the ceph_volume_client. https://github.com/ceph/ceph/blob/master/qa/tasks/cephfs/test_volume_client.py#L733 shows contents of a auth meta file.

This could be saved for future work.